solr/core/src/test/org/apache/solr/handler/TestHttpRequestId.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.handler;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.apache.solr.SolrJettyTestBase;
+import org.apache.solr.client.solrj.impl.Http2SolrClient;
+import org.apache.solr.client.solrj.impl.HttpListenerFactory.RequestResponseListener;
+import org.apache.solr.client.solrj.request.SolrPing;
+import org.apache.solr.client.solrj.util.AsyncListener;
+import org.apache.solr.common.util.ExecutorUtil;
+import org.apache.solr.common.util.NamedList;
+import org.apache.solr.common.util.SolrNamedThreadFactory;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.api.Result;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.slf4j.MDC;
+
+public class TestHttpRequestId extends SolrJettyTestBase {
```

Comment:
Your test approach is a bit invasive... uses the internals of Http2 client.  It's okay but I was hoping you would have more of an integration test that like TestRequestId, which is why I referred you to that test.  It does not depend on internal details of SolrClients etc.  I love integration tests as it's more realistic, doesn't assume internals, and also makes it easier to change internals at some point.  Since I now understand you observed specifically an issue with distributed search because of HttpShardHandler's use of that, maybe you could even use a simple query that uses that under the hood?

Reply:
I don't think your example does covers this scenario out of the box. this bug is about MDC being present across the threads running the request. there are no logs at this level, unless I add some specifically for this test, something trace level, but that feels awkward.

Reply:
I'm confused.  You are basically saying that the improvements here are un-observable?  I wonder what motivated this dubious improvement then -- something custom you have?
FWIW I don't think it'd be awkward if you added a trace log.  And FYI it's easy to configure a log level for just one test via the `LogLevel` annotation -- a wonderful thing.

Reply:
> You are basically saying that the improvements here are un-observable? I wonder what motivated this dubious improvement then -- something custom you have?

there might be a small misunderstanding here. It's not am improvement I am chasing here, it's a bug, a regression in Solr 9 compared to Solr 8. and I don't have anything custom, this is inside the http2 client as it is used by the HttpShardHandler.
I am trying to setup LogLevel + LogListener in a SolrCloudTestCase, see if I can get something better looking.

Reply:
in the end it was SolrJettyTestBase not the SolrCloudTestCase as I mentioned before. please take a look at this version of the test: it uses logs to register context and verifies existence of MDC context.